### PR TITLE
feat(Optimizer): Allow for batch ask requests

### PR DESCRIPTION
**Pull request by eddiebergman**
This PR extends the `.ask()` of Optimizers to include an integer `ask(n: int | None = None)` where the default behaviour is to get a single configuration, while giving an `n` returns an iterator of configurations.

By default, optimizers which work in a sequential manner can simply call to them selves in a repeated fashion:

```python
if n is not None:
    return (self.ask(n=None) for _ range(n))
```

This PR is more aimed at supporting optimizers which support batched `ask`s. This can provide both theoretical improvements but also more optimized improvements. For example, if there is an evolutionary algorithm, computing `n` children can be faster than computing `1` child `n` times.

TODO:
* Currently every optimizer implemented just defaults to the sequential ask mechanism provided in the snippet above. We need to investigate if Optimizers support the batching mechanism and implement it if so. I've reached out to SMAC and NEPS. #225 
* Document it's usage #226 

---

This PR was done as a stepping stone towards another feature in progress I wanted to have this `n` positional parameter in a `PipelineOptimizer` be reserved, even though no optimizer has this batching implemented yet.

### CODE REVIEW

File: src/amltk/optimization/optimizers/smac.py [hunk 1]
```diff
@@ -249,13 +249,31 @@ def create(
...
+
+    @overload
+    def ask(self, n: None = None) -> Trial[SMACTrialInfo]:
+        ...
+
     @override
-    def ask(self) -> Trial[SMACTrialInfo]:
+    def ask(
+        self,
+        n: int | None = None,
```
Comment (LennartPurucker) [inline src/amltk/optimization/optimizers/smac.py:263]: Mhm, I think it might be unintuitive to make the default `None`. I would rather have it be `1`. 
I know you liked had to do it this way for the recursion to work but maybe it would be cleaner to just use a new `_ask` function for that and make the public-facing interface use `1 <= n <= +inf`.

File: tests/optimizers/test_optimizers.py [hunk 2]
```diff
@@ -128,3 +129,13 @@ def test_report_failure(optimizer: Optimizer):
...
+
+
+@parametrize_with_cases("optimizer", cases=".", prefix="opt_")
+def test_batched_ask_generates_unique_configs(optimizer: Optimizer):
+    """Test that batched ask generates unique configs."""
+    # NOTE: This was tested with up to 100, at least from SMAC and Optuna.
+    # It was quite slow for smac so I've reduced it to 10.
+    # This is not a hard requirement of optimizers (maybe it should be?)
+    batch = optimizer.ask(10)
+    assert all_unique(batch)
```
Comment (LennartPurucker) [inline tests/optimizers/test_optimizers.py:141]: This does not test that it also returns 10 trials, right?

**Review by LennartPurucker — CHANGES_REQUESTED** [review]
See comments below, otherwise LGTM

File: src/amltk/optimization/optimizers/smac.py (hunk 1 above)
Comment (eddiebergman) [reply src/amltk/optimization/optimizers/smac.py:263]: The reason was actually for typing. If you use an `int`, you get a `list` of that size, otherwise you get a single trial.

```python
def f(n: int = 1) -> str | list[str]:
    if n == 1:
		return "hello"
	return ["hello"] * n

n = random.integer()
for message in f(n):
    print(message)  # If you get unlucky and have `1`, this would print 'h', 'e', 'l', 'l', 'o'
```

The `None` thing is basically saying that you get a single item, if you pass an integer, you get a list of that size.

Typing becomes a nightmare if you try to do it based on `Literal[1]`.

---

tldr; I won't do this change as it can cause unpredictable behaviour and doesn't offer much benefit

File: tests/optimizers/test_optimizers.py (hunk 2 above)
Comment (eddiebergman) [reply tests/optimizers/test_optimizers.py:141]: Ohh good point, should definitely include that!
